diff --git a/node_modules/@kalashshah/restapi/src/lib/chat/helpers/payloadHelper.ts b/node_modules/@kalashshah/restapi/src/lib/chat/helpers/payloadHelper.ts
index b322a8e..72db613 100644
--- a/node_modules/@kalashshah/restapi/src/lib/chat/helpers/payloadHelper.ts
+++ b/node_modules/@kalashshah/restapi/src/lib/chat/helpers/payloadHelper.ts
@@ -345,8 +345,6 @@ export const updateGroupPayload = (
     scheduleAt: scheduleAt,
     scheduleEnd: scheduleEnd,
     status: status,
-    ...(meta !== undefined && { meta: meta }),
-    ...(rules !== undefined && { rules: rules }),
   };
   return body;
 };
@@ -377,7 +375,7 @@ export const getAdminsList = (
       )
     : [];
 
-  const adminList = [...adminsFromMembers, ...adminsFromPendingMembers];
+  const adminList = []//[...adminsFromMembers, ...adminsFromPendingMembers];
   return adminList;
 };
 
@@ -405,7 +403,7 @@ export const getSpaceAdminsList = (
       )
     : [];
 
-  const adminList = [...adminsFromMembers, ...adminsFromPendingMembers];
+  const adminList = []//[...adminsFromMembers, ...adminsFromPendingMembers];
   return adminList;
 };
 
@@ -429,7 +427,7 @@ export const getMembersList = (
     image: string;
   }[]
 ): Array<string> => {
-  const allMembers = [...(members || []), ...(pendingMembers || [])];
+  const allMembers = []//[...(members || []), ...(pendingMembers || [])];
   return convertToWalletAddressList(allMembers);
 };
 
@@ -447,6 +445,6 @@ export const getSpacesMembersList = (
     image: string;
   }[]
 ): Array<string> => {
-  const allMembers = [...(members || []), ...(pendingMembers || [])];
+  const allMembers = []//[...(members || []), ...(pendingMembers || [])];
   return convertToWalletAddressList(allMembers);
 };
