diff --git a/node_modules/react-native/index.js b/node_modules/react-native/index.js
index d59ba34..c378e45 100644
--- a/node_modules/react-native/index.js
+++ b/node_modules/react-native/index.js
@@ -456,7 +456,7 @@ module.exports = {
     );
   },
   get ViewPropTypes(): $FlowFixMe {
-    invariant(
+    console.warn(
       false,
       'ViewPropTypes has been removed from React Native. Migrate to ' +
         "ViewPropTypes exported from 'deprecated-react-native-prop-types'.",
