// @unstoppabledomains/resolution/build/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var Resolution_1 = require("./Resolution");
exports.Resolution = Resolution_1.Resolution;
exports.default = Resolution_1.Resolution;
__export(require("./types/publicTypes"));
var DnsUtils_1 = require("./utils/DnsUtils");
exports.DnsUtils = DnsUtils_1.default;
var resolutionError_1 = require("./errors/resolutionError");
exports.ResolutionError = resolutionError_1.ResolutionError;
exports.ResolutionErrorCode = resolutionError_1.ResolutionErrorCode;
var configurationError_1 = require("./errors/configurationError");
exports.ConfigurationError = configurationError_1.ConfigurationError;
exports.ConfigurationErrorCode = configurationError_1.ConfigurationErrorCode;
var dnsRecordsError_1 = require("./errors/dnsRecordsError");
exports.DnsRecordsError = dnsRecordsError_1.DnsRecordsError;
exports.DnsRecordsErrorCode = dnsRecordsError_1.DnsRecordsErrorCode;
var Eip1993Factories_1 = require("./utils/Eip1993Factories");
exports.Eip1993Factories = Eip1993Factories_1.Eip1993Factories;
